### PR TITLE
feat(helm): add lifecycle hooks to alloy container

### LINKED ISSUE #1140: Container lifecycle hook support in helm chart
### Request

Container lifecycle hook support is needed as an option in the helm chart to handle graceful pod termination when deployed behind an AWS Application Loadbalancer.

### Use case

When using Grafana Alloy in a Kubernetes deployment and with an Ingress provisioned by the AWS ALB Ingress controller, it is required that you use a PreStop lifecycle hook. This is needed to pause the termination of the pod/container to allow the ALB to drain its connection before terminating the pod. Otherwise, an early pod termination may lead to 5xx errors being sent to the client.

This can be mitigated by configuring the deregistration delay on the target group (using ingress annotations) and adding a PreStop hook to pause the termination of the pod and allow it to be deregistered from the target group before it gets terminated.

See AWS ALB Controller issue: https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1064

### LINKED ISSUE #1673: add lifecycle on pod_template
### Request

We running alloy in daemonset to collect pod logs. 
We use AWS spot node, it often was  removed. But the alloy pods recycled faster then other pods, it will cause some log didn't send to loki.

Can you help add `lifecycle` and `terminationGracePeriodSeconds` to pod_template?
```
lifecycle:
  preStop:
    exec:
      command: ["sleep", "30"]
```

### Use case

Can you help add `lifecycle` and `terminationGracePeriodSeconds` to pod_template?
```
terminationGracePeriodSeconds: 60
lifecycle:
  preStop:
    exec:
      command: ["sleep", "30"]
```

### PR DESCRIPTION
**Pull request by etiennep**
#### Which issue(s) this PR fixes

Fixes #1140
Fixes #1673

Adds lifecycle hooks to the alloy container to allow graceful target de-registration when using with AWS LoadBalancer Controller ingress.

#### PR Checklist

- [X] CHANGELOG.md updated
- [X] Documentation added
- [X] Tests updated

### CODE REVIEW

**Comment by mattdurham** [comment]
This seems reasonable but looping in @petewall for any thoughts.

**Comment by petewall** [comment]
> This seems reasonable but looping in @petewall for any thoughts.

Left a few comments. All in all, looks good to me.

File: operations/helm/charts/alloy/values.yaml [hunk 1]
```diff
@@ -183,6 +191,9 @@ controller:
   # -- Configures the DNS policy for the pod. https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy
   dnsPolicy: ClusterFirst
 
+  # -- Termination grace period in seconds for the Grafana Alloy pods.
+  terminationGracePeriodSeconds: null
```
[the remark this answers is not in the document]
Comment (etiennep) [reply operations/helm/charts/alloy/values.yaml:195]: According to Kubernetes documentation, the default is 30s. https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/

I will add a note to the comment.

**Review by mattdurham — APPROVED** [review]
LGTM

**Comment by mattdurham** [comment]
@clayton-cornell for review of docs.

**Review by clayton-cornell — APPROVED** [review]
LGTM

**Comment by mattdurham** [comment]
Seeing some helm linting errors.

**Comment by etiennep** [comment]
I think the linting issues should be fixed now @mattdurham 

Thank you.

**Pull request merged**